sentry/src/main/java/io/sentry/SentryOptions.java

```diff
@@ -1,60 +1,311 @@
 package io.sentry;
 
+import static java.util.Arrays.asList;
+import static java.util.Collections.emptyList;
+import static java.util.Collections.singletonList;
+import static java.util.Objects.requireNonNull;
+
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import io.sentry.config.CompoundResourceLoader;
+import io.sentry.config.ContextClassLoaderResourceLoader;
+import io.sentry.config.FileResourceLoader;
+import io.sentry.config.Lookup;
 import io.sentry.config.ResourceLoader;
+import io.sentry.config.location.CompoundResourceLocator;
+import io.sentry.config.location.ConfigurationResourceLocator;
+import io.sentry.config.location.EnvironmentBasedLocator;
+import io.sentry.config.location.StaticFileLocator;
+import io.sentry.config.location.SystemPropertiesBasedLocator;
+import io.sentry.config.provider.CompoundConfigurationProvider;
+import io.sentry.config.provider.ConfigurationProvider;
+import io.sentry.config.provider.EnvironmentConfigurationProvider;
+import io.sentry.config.provider.JndiConfigurationProvider;
+import io.sentry.config.provider.JndiSupport;
+import io.sentry.config.provider.LocatorBasedConfigurationProvider;
+import io.sentry.config.provider.SystemPropertiesConfigurationProvider;
+import io.sentry.dsn.Dsn;
+import io.sentry.util.Nullable;
+import io.sentry.util.Util;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * SentryOptions used to configure the Sentry SDK.
  */
-public class SentryOptions {
-    private SentryClientFactory sentryClientFactory;
-    private String dsn;
-    private ResourceLoader resourceLoader;
+public final class SentryOptions {
+    private static final Logger logger = LoggerFactory.getLogger(SentryOptions.class);
+
+    private final Lookup lookup;
+    private final Dsn dsn;
+    private final SentryClientFactory clientFactory;
+
+    /**
+     * Creates a new instance of the options using the provided parameters.
+     *
+     * @param lookup        the lookup to locate the configuration with
+     * @param dsn           the dsn to use
+     * @param clientFactory the client factory to use
+     * @throws NullPointerException if any of the parameters is null
+     */
+    public SentryOptions(Lookup lookup, Dsn dsn, SentryClientFactory clientFactory) {
+        this.lookup = requireNonNull(lookup, "lookup");
+        this.dsn = requireNonNull(dsn, "dsn");
```

Comment:
An empty dsn is a valid use-case. We just keep the SDK disabled. This is a documented behavior we have in all Sentry SDKs.
Also having a public setter for dsn allows users read their dsn after the options was instantiated. Say a framework that creates the options instance for you and binds values from some configuration source you define.